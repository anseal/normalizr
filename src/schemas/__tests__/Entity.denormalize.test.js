// eslint-env jest
import { denormalize, schema } from '../../'

describe('EntitySchema denormalization', () => {
	test('denormalizes an entity', () => {
		expect(true).toBe(true)
		const mySchema = new schema.Entity('tacos')
		const entities = {
			tacos: {
				1: { id: 1, type: 'foo' },
			},
		}
		expect(denormalize(1, mySchema, entities)).toMatchSnapshot()
	})

	test('denormalizes deep entities', () => {
		const foodSchema = new schema.Entity('foods')
		const menuSchema = new schema.Entity('menus', {
			food: foodSchema,
		})

		const entities = {
			menus: {
				1: { id: 1, food: 1 },
				2: { id: 2 },
			},
			foods: {
				1: { id: 1 },
			},
		}

		expect(denormalize(1, menuSchema, entities)).toMatchSnapshot()
		expect(denormalize(2, menuSchema, entities)).toMatchSnapshot()
	})

	test('denormalizes to undefined for missing data', () => {
		const foodSchema = new schema.Entity('foods')
		const menuSchema = new schema.Entity('menus', {
			food: foodSchema,
		})

		const entities = {
			menus: {
				1: { id: 1, food: 2 },
			},
			foods: {
				1: { id: 1 },
			},
		}

		expect(denormalize(1, menuSchema, entities)).toMatchSnapshot()
		expect(denormalize(2, menuSchema, entities)).toMatchSnapshot()
	})

	test('can denormalize already partially denormalized data', () => {
		const foodSchema = new schema.Entity('foods')
		const menuSchema = new schema.Entity('menus', {
			food: foodSchema,
		})

		const entities = {
			menus: {
				1: { id: 1, food: { id: 1 } },
			},
			foods: {
				1: { id: 1 },
			},
		}

		expect(denormalize(1, menuSchema, entities)).toMatchSnapshot()
	})

	test('denormalizes recursive dependencies', () => {
		const user = new schema.Entity('users')
		const report = new schema.Entity('reports')

		user.define({
			reports: [report],
		})
		report.define({
			draftedBy: user,
			publishedBy: user,
		})

		const entities = {
			reports: {
				123: {
					id: '123',
					title: 'Weekly report',
					draftedBy: '456',
					publishedBy: '456',
				},
			},
			users: {
				456: {
					id: '456',
					role: 'manager',
					reports: ['123'],
				},
			},
		}
		expect(denormalize('123', report, entities)).toMatchSnapshot()
		expect(denormalize('456', user, entities)).toMatchSnapshot()
	})

	test('denormalizes entities with referential equality', () => {
		const user = new schema.Entity('users')
		const report = new schema.Entity('reports')

		user.define({
			reports: [report],
		})
		report.define({
			draftedBy: user,
			publishedBy: user,
		})

		const entities = {
			reports: {
				123: {
					id: '123',
					title: 'Weekly report',
					draftedBy: '456',
					publishedBy: '456',
				},
			},
			users: {
				456: {
					id: '456',
					role: 'manager',
					reports: ['123'],
				},
			},
		}

		const denormalizedReport = denormalize('123', report, entities)

		expect(denormalizedReport).toBe(denormalizedReport.draftedBy.reports[0])
		expect(denormalizedReport.publishedBy).toBe(denormalizedReport.draftedBy)

		// NOTE: Given how immutable data works, referential equality can't be
		// maintained with nested denormalization.
	})

	test('denormalizes with fallback strategy', () => {
		const user = new schema.Entity(
			'users',
			{},
			{
				idAttribute: 'userId',
				fallbackStrategy: (id, schema) => ({
					[schema.idAttribute]: id,
					name: 'John Doe',
				}),
			}
		)
		const report = new schema.Entity('reports', {
			draftedBy: user,
			publishedBy: user,
		})

		const entities = {
			reports: {
				123: {
					id: '123',
					title: 'Weekly report',
					draftedBy: '456',
					publishedBy: '456',
				},
			},
			users: {},
		}

		const denormalizedReport = denormalize('123', report, entities)

		expect(denormalizedReport.publishedBy).toBe(denormalizedReport.draftedBy)
		expect(denormalizedReport.publishedBy.name).toBe('John Doe')
		expect(denormalizedReport.publishedBy.userId).toBe('456')
		//
	})
})
