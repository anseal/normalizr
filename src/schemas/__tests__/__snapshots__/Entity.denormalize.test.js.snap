// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitySchema denormalization can denormalize already partially denormalized data 1`] = `
Object {
  "food": Object {
    "id": 1,
  },
  "id": 1,
}
`;

exports[`EntitySchema denormalization denormalizes an entity 1`] = `
Object {
  "id": 1,
  "type": "foo",
}
`;

exports[`EntitySchema denormalization denormalizes deep entities 1`] = `
Object {
  "food": Object {
    "id": 1,
  },
  "id": 1,
}
`;

exports[`EntitySchema denormalization denormalizes deep entities 2`] = `
Object {
  "id": 2,
}
`;

exports[`EntitySchema denormalization denormalizes recursive dependencies 1`] = `
Object {
  "draftedBy": Object {
    "id": "456",
    "reports": Array [
      [Circular],
    ],
    "role": "manager",
  },
  "id": "123",
  "publishedBy": Object {
    "id": "456",
    "reports": Array [
      [Circular],
    ],
    "role": "manager",
  },
  "title": "Weekly report",
}
`;

exports[`EntitySchema denormalization denormalizes recursive dependencies 2`] = `
Object {
  "id": "456",
  "reports": Array [
    Object {
      "draftedBy": [Circular],
      "id": "123",
      "publishedBy": [Circular],
      "title": "Weekly report",
    },
  ],
  "role": "manager",
}
`;

exports[`EntitySchema denormalization denormalizes to undefined for missing data 1`] = `
Object {
  "food": undefined,
  "id": 1,
}
`;

exports[`EntitySchema denormalization denormalizes to undefined for missing data 2`] = `undefined`;
